### PR TITLE
Closes #211 | Implement dataloader for SEAHORSE

### LINKED ISSUE #211: Create dataset loader for Seahorse
Dataloader name: `seahorse/seahorse.py`
DataCatalogue: http://seacrowd.github.io/seacrowd-catalogue/card.html?seahorse

|     Dataset| seahorse |
|-------------|---|
| Description | SEAHORSE is a dataset for multilingual, multifaceted summarization evaluation. It consists of 96K summaries with human ratings along 6 quality dimensions: comprehensibility, repetition, grammar, attribution, main idea(s), and conciseness, covering 6 languages, 9 systems and 4 datasets. |
| Subsets | - |
| Languages | vie |
| Tasks | Summarization |
| License | Creative Commons Attribution 4.0 (cc-by-4.0) |
| Homepage | https://storage.googleapis.com/seahorse-public/seahorse_data.zip |
| HF URL | - |
| Paper URL | https://aclanthology.org/2023.emnlp-main.584 |

### PR DESCRIPTION
**Pull request by chenxwh**
Closes #211

### Checkbox
- [x] Confirm that this PR is linked to the dataset issue.
- [x] Create the dataloader script `seacrowd/sea_datasets/my_dataset/my_dataset.py` (please use only lowercase and underscore for dataset naming).
- [x] Provide values for the `_CITATION`, `_DATASETNAME`, `_DESCRIPTION`, `_HOMEPAGE`, `_LICENSE`, `_URLs`, `_SUPPORTED_TASKS`, `_SOURCE_VERSION`, and `_SEACROWD_VERSION` variables.
- [x] Implement `_info()`, `_split_generators()` and `_generate_examples()` in dataloader script.
- [x] Make sure that the `BUILDER_CONFIGS` class attribute is a list with at least one `SEACrowdConfig` for the source schema and one for a seacrowd schema.
- [x] Confirm dataloader script works with `datasets.load_dataset` function.
- [x] Confirm that your dataloader script passes the test suite run with `python -m tests.test_seacrowd seacrowd/sea_datasets/<my_dataset>/<my_dataset>.py`.
- [x] If my dataset is local, I have provided an output of the unit-tests in the PR (please copy paste). This is OPTIONAL for public datasets, as we can test these without access to the data files.

### CODE REVIEW

File: requirements.txt
Comment (fhudi) [inline requirements.txt:28]: There are 2 considerations on my end by changing the requirements for the whole project:

1. Use of exact match specifier.
I think this should be discouraged for a better compatibility. In this context my reasonings are:
    - Are these exact version numbers required by google-research's SEAHORSE?
    - Does the specifier indicate that other versions would break?
    - The versions are relatively new release, i.e.: conda-forge version of `tensorflow-datasets` is only [4.8.3](https://anaconda.org/conda-forge/tensorflow-datasets) (as of 27 Feb '24).

2. The necessity of depending on the libraries
I did not give any suggestion on point 1, sorry for that, and it is related to this point.
Some questions I would like to raise:
    - Is it necessary to depend on `tensorflow`?
      IMHO, SEACrowd should not directly depend on any DL Framework.
    - Is there any work-around for not depending on `tensorflow-datasets` to access the dataset?
      i.e.: download the raw dataset and extract it ourselves?
      (Although, I think dependancy on `tensorflow_datasets` is still reasonable)

**Review by fhudi — CHANGES_REQUESTED** [review]
Please change to access the data directly rather than other's HF-Datasets repo

File: requirements.txt
Comment (chenxwh) [reply requirements.txt]: updated accordingly

**Comment by chenxwh** [comment]
updated accordingly

File: seacrowd/sea_datasets/seahorse/seahorse.py
Comment (chenxwh) [reply seacrowd/sea_datasets/seahorse/seahorse.py:66]: following https://github.com/google-research-datasets/seahorse?tab=readme-ov-file#retrieving-articles-from-gem, what is your suggestion?

File: seacrowd/sea_datasets/seahorse/seahorse.py
Comment (chenxwh) [reply seacrowd/sea_datasets/seahorse/seahorse.py:83]: so save in an additional `dict` first?

Comment (fhudi) [inline seacrowd/sea_datasets/seahorse/seahorse.py:83]:
```suggestion
    if "wiki_lingua" in gem_id:
        data = _WIKILINGUAL_DATA
    elif "xlsum" in gem_id:
        data = _XLSUM_DATA
    else:
        raise AssertionError("gem_id should either from wiki_lingua or xlsum.")
    return data[split if split == "test" else "validation"][gem_id]
```

File: seacrowd/sea_datasets/seahorse/seahorse.py
Comment (fhudi) [inline seacrowd/sea_datasets/seahorse/seahorse.py:66]:
```suggestion
    df = tfds.as_dataframe(ds, info)
    return dict(zip(*[df[col].str.decode("utf8") for col in ["gem_id", "source"]]))
```

File: seacrowd/sea_datasets/seahorse/seahorse.py
Comment (fhudi) [inline seacrowd/sea_datasets/seahorse/seahorse.py:68]:
```suggestion
def get_xlsum_data(lang, split):
    df = datasets.load_dataset("GEM/xlsum", lang)
    return {item["gem_id"]: item["text"] for item in df[split]}

```

File: seacrowd/sea_datasets/seahorse/seahorse.py
Comment (fhudi) [inline seacrowd/sea_datasets/seahorse/seahorse.py:71]:
```suggestion
_XLSUM_DATA = {split: get_xlsum_data("vietnamese", split) for split in ["test", "validation"]}
```

**Review by fhudi — CHANGES_REQUESTED** [review]
Please check the comments for the additional id-to-text mapping along with the removal the superfluous whole dataset storage and transformation for each sample case.

P.S.: you can try committing the 4 changes ([a](https://github.com/SEACrowd/seacrowd-datahub/pull/407#discussion_r1506321973), [b](https://github.com/SEACrowd/seacrowd-datahub/pull/407#discussion_r1506323328), [c](https://github.com/SEACrowd/seacrowd-datahub/pull/407#discussion_r1506326119), [d](https://github.com/SEACrowd/seacrowd-datahub/pull/407#discussion_r1506326603)) and run the `python -m tests.test_seacrowd xxx`.

File: seacrowd/sea_datasets/seahorse/seahorse.py
Comment (fhudi) [reply seacrowd/sea_datasets/seahorse/seahorse.py:83]: Yes.
For this particular part, please refer to the [latest change request](https://github.com/SEACrowd/seacrowd-datahub/pull/407#pullrequestreview-1906855080).

File: seacrowd/sea_datasets/seahorse/seahorse.py
Comment (fhudi) [reply seacrowd/sea_datasets/seahorse/seahorse.py:66]: I suggest to directly access the dataset instead, both the wikilingua and xlsum.
You can refer to [the original paper](https://aclanthology.org/2020.findings-emnlp.360/), and you can refer to the [huggingface's implementation](https://huggingface.co/datasets/GEM/wiki_lingua/blob/main/wiki_lingua.py) on how to implement this dataloader.
However, I would like to check whether adding the `tensorflow-dataset` library is alright.

---
@holylovenia @SamuelCahyawijaya Sorry for summoning you guys, but wdyt?

**Comment by chenxwh** [comment]
@fhudi thanks for the comments and suggestions, I was trying to incorporate the changes but it does not look like one can use `tensorflow-data` or `tfds-nightly` without `tensorflow`, it is mentioned here https://www.tensorflow.org/datasets/tfless_tfds that the first time to prepare for the datasets `tensorflow` is still required

File: seacrowd/sea_datasets/seahorse/seahorse.py
Comment (chenxwh) [reply seacrowd/sea_datasets/seahorse/seahorse.py]: I doubt the `seahorse` version of Wikilingua is the same as the original one. It is the version that is used for the GEM shared task. BTW `tfds` cannot be used without `tensorflow` it seems

**Comment by fhudi** [comment]
> @fhudi thanks for the comments and suggestions, I was trying to incorporate the changes but it does not look like one can use `tensorflow-data` or `tfds-nightly` without `tensorflow`, it is mentioned here https://www.tensorflow.org/datasets/tfless_tfds that the first time to prepare for the datasets `tensorflow` is still required

@chenxwh 
Yeah, that's why I requested for direct access to the dataset files rather than through other's DataLoaders..

**Comment by chenxwh** [comment]
Though I don't think the original Wikiligua is the same as what Seahorse uses that is from the gem task

**Comment by fhudi** [comment]
> what is the harm of having `tensorflow`?

It is not as extreme to say it as a "harm".
Normally, the dataset is independent from the Neural Network Frameworks, like pytorch and tensorflow.

@SamuelCahyawijaya @holylovenia Please advise, and CMIIW.

**Comment by holylovenia** [comment]
> > what is the harm of having `tensorflow`?
> 
> It is not as extreme to say it as a "harm". Normally, the dataset is independent from the Neural Network Frameworks, like pytorch and tensorflow.
> 
> @SamuelCahyawijaya @holylovenia Please advise, and CMIIW.

I'm going to have a discussion with @SamuelCahyawijaya and @sabilmakbar in a few. I'll get back to you on this.

**Comment by chenxwh** [comment]
Of course happy to bypass it if we can, although I am not sure how - if the "mapping" itself requires `tfds`, at least the very first time to obtain it.

**Comment by holylovenia** [comment]
@chenxwh Is there any reason for not using `pickle` just like [here](https://huggingface.co/datasets/GEM/wiki_lingua/blob/main/wiki_lingua.py)?

cc: @fhudi @sabilmakbar @SamuelCahyawijaya

File: seacrowd/sea_datasets/seahorse/seahorse.py
Comment (holylovenia) [inline seacrowd/sea_datasets/seahorse/seahorse.py:65]: We also tried to look for alternatives other than using `tfds`, but couldn't find them. We also found that the wiki_lingua in HF and `tfds` are different in numbers. It could be that the hf dataset is GEMv2 (not GEMv1) and has been re-split. To ensure reproducibility, it would be better to proceed with `tfds` and modify the dataloader to require `tfds` with try-catch, @chenxwh.

Example:
```
try:
    import yourModule
except ImportError:
    raise Exception(“…”)
```

cc: @fhudi @SamuelCahyawijaya @sabilmakbar

Comment (chenxwh) [reply seacrowd/sea_datasets/seahorse/seahorse.py:65]: @holylovenia thanks for the comment, you mean do not add `tensorflow-dataset` in the requirements.txt either? but try import `tfds` in line 65? 
And `tensorflow`? we don't need to import it explicitly in the script but it is required by `tfds`

Comment (holylovenia) [reply seacrowd/sea_datasets/seahorse/seahorse.py:65]: Ah yeah, no need to add the additional packages that you need for this specific dataloader, e.g., `tensorflow`, to the requirements.txt. We can just use try-catch to ensure the user has it if they specifically want to use this dataloader.

Comment (chenxwh) [reply seacrowd/sea_datasets/seahorse/seahorse.py]: Thanks @holylovenia, updated now, put more general exception to catch the error.

**Comment by sabilmakbar** [comment]
[rewriting as separate comment for better visibility]

Hi @chenxwh, after checking on `tfds` implementation that uses HF namespace[1], it will execute the tfds dataloader construction as follows:
1. It parse the name config of `huggingface:gem/wiki_lingua_vietnamese_vi` to check the data source, name, and config (if provided)
2. call the HF Constructor on TFDS[2]
3. And uses `as_dataset` method from HF Class [3], similar to what `load_dataset` does in HF dataset.
4. For the info on TFDS, it will call this method on HF Constructor [4], but I think it's unnecessary if we can create an equivalence to the HF `datasets.load_dataset`

Hence, the hypothetical equivalent code should be:
```python
ds = datasets.load_dataset("gem", name="wiki_lingua_vietnamese_vi")
df = ds.to_pandas()
```

Can you try to check the above suggestion on your `get_wikilingual_data` fn and see if the data exactly matches? Since from @holylovenia's attempt, we uses `wiki_lingua` dataset that's slightly different from what the Seabench uses.

cc @holylovenia @SamuelCahyawijaya 

[1] https://github.com/tensorflow/datasets/blob/ab25353173afa5fa01d03759a5e482c82d4fd889/tensorflow_datasets/core/load.py#L185
[2] https://github.com/tensorflow/datasets/blob/ab25353173afa5fa01d03759a5e482c82d4fd889/tensorflow_datasets/core/dataset_builders/huggingface_dataset_builder.py#L437
[3] https://github.com/tensorflow/datasets/blob/ab25353173afa5fa01d03759a5e482c82d4fd889/tensorflow_datasets/core/dataset_builders/huggingface_dataset_builder.py#L390
[4] https://github.com/tensorflow/datasets/blob/ab25353173afa5fa01d03759a5e482c82d4fd889/tensorflow_datasets/core/dataset_builders/huggingface_dataset_builder.py#L403

**Comment by chenxwh** [comment]
> [rewriting as separate comment for better visibility]
> 
> 
> 
> Hi @chenxwh, after checking on `tfds` implementation that uses HF namespace[1], it will execute the tfds dataloader construction as follows:
> 
> 1. It parse the name config of `huggingface:gem/wiki_lingua_vietnamese_vi` to check the data source, name, and config (if provided)
> 
> 2. call the HF Constructor on TFDS[2]
> 
> 3. And uses `as_dataset` method from HF Class [3], similar to what `load_dataset` does in HF dataset.
> 
> 4. For the info on TFDS, it will call this method on HF Constructor [4], but I think it's unnecessary if we can create an equivalence to the HF `datasets.load_dataset`
> 
> 
> 
> Hence, the hypothetical equivalent code should be:
> 
> ```python
> 
> ds = datasets.load_dataset("gem", name="wiki_lingua_vietnamese_vi", split=split) # the value of "spit" is inferred from the `get_wikilingual_data` accordingly
> 
> df = ds.to_pandas()
> 
> ```
> 
> 
> 
> Can you try to check the above suggestion on your `get_wikilingual_data` fn and see if the data exactly matches? Since from @holylovenia's attempt, we uses `wiki_lingua` dataset that's slightly different from what the Seabench uses.
> 
> 
> 
> cc @holylovenia @SamuelCahyawijaya 
> 
> 
> 
> [1] https://github.com/tensorflow/datasets/blob/ab25353173afa5fa01d03759a5e482c82d4fd889/tensorflow_datasets/core/load.py#L185
> 
> [2] https://github.com/tensorflow/datasets/blob/ab25353173afa5fa01d03759a5e482c82d4fd889/tensorflow_datasets/core/dataset_builders/huggingface_dataset_builder.py#L437
> 
> [3] https://github.com/tensorflow/datasets/blob/ab25353173afa5fa01d03759a5e482c82d4fd889/tensorflow_datasets/core/dataset_builders/huggingface_dataset_builder.py#L390
> 
> [4] https://github.com/tensorflow/datasets/blob/ab25353173afa5fa01d03759a5e482c82d4fd889/tensorflow_datasets/core/dataset_builders/huggingface_dataset_builder.py#L403
> 
> 
> 
> Also, the GEM Wikilingua Dataset on [this URL](https://www.tensorflow.org/datasets/catalog/gem) matches with the HF Pointer that I provided in the above snippet

See https://github.com/google-research-datasets/seahorse?tab=readme-ov-file#retrieving-articles-from-gem

Seahorse does not use GEM/wikingual on hf, as specified pointed out, hence needs to be loaded differently than the other two datasets, where the "earlier" version of the mapping is stored in tfds

**Comment by holylovenia** [comment]
> See https://github.com/google-research-datasets/seahorse?tab=readme-ov-file#retrieving-articles-from-gem
> 
> Seahorse does not use GEM/wikingual on hf, as specified pointed out, hence needs to be loaded differently than the other two datasets, where the "earlier" version of the mapping is stored in tfds

Hi @chenxwh, I think you misunderstood @sabilmakbar's suggestion.

We know that the data loaded by `tfds.load(f'huggingface:gem/wiki_lingua_{lang}'` following [this guide](https://github.com/google-research-datasets/seahorse?tab=readme-ov-file#retrieving-articles-from-gem) is different from HF's `datasets.load("GEM/wiki_lingua") [here](https://huggingface.co/datasets/GEM/wiki_lingua/viewer/vi). For this reason, @sabilmakbar inspected how `tfds.load(...)` that uses the HF namespace, e.g., `"huggingface:gem/wiki_lingua_vi"`, works, thus we can replicate how `tfds.load(...)` works without using both the `tfds` and `tensorflow` library.

Based on the flow he explained in the previous comment, @sabilmakbar created a non-`tfds` code that replicates the `tfds.load(...)` result:

```
ds = datasets.load_dataset("gem", name="wiki_lingua_vietnamese_vi", split=split) # the value of "spit" is inferred from the `get_wikilingual_data` accordingly
df = ds.to_pandas()
```

I hope this clears up things, @chenxwh. For more details, you can refer to his comment.

Correct me if I missed anything, @sabilmakbar.

**Comment by chenxwh** [comment]
Ah I see thanks for the clarification @holylovenia, it's now updated accordingly

**Comment by holylovenia** [comment]
> Ah I see thanks for the clarification @holylovenia, it's now updated accordingly

Thanks for understanding, @chenxwh! Also thanks for finding this workaround, @sabilmakbar. 👍 

Feel free to proceed with the reviewing process, @fhudi @MJonibek.

**Review by fhudi — APPROVED** [review]
LGTM, thanks all for the hard-work.
As far as I'm concerned, now it has no superfluous stuffs, and the code is cleaner and neater.
(and passed the reviewing checklists 😅)

Thanks for complying, @chenxwh.
Thanks for the clarifications and recommendations, @holylovenia @sabilmakbar.

---

Let's wait for the review from @MJonibek.

**Review by MJonibek — APPROVED** [review]
@fhudi @chenxwh Great work done, thank you. Checked, LGTM

**Pull request merged**
